src/LuaUtils.cpp

```diff
@@ -171,127 +171,363 @@ static bool _import_core(lua_State *L, const std::string &importName)
+	return caller;
+}
+
+/**
+* Finds field in table.
+* Requires a existing table in the stack at index -1.
+* Returns true if field exists and not nil, puts value to lua stack.
+*/
+static bool get_cached(lua_State *L, const std::string& name)
+{
+	LUA_DEBUG_START(L);
 	assert(lua_istable(L, -1));
 
-	// first check the cache for the explicit name (its there if it was
-	// previously imported from core)
-	std::string cacheName = importName;
-	lua_getfield(L, -1, cacheName.c_str());
-	if (!lua_isnil(L, -1)) {
-		// found it, use it
-		DEBUG_PRINTF("import: found cached %s\n", cacheName.c_str());
-		lua_remove(L, -2);
-		return 1;
+	// try to get it
+	lua_getfield(L, -1, name.c_str());
+
+	// if got cached field
+	bool isCached = !lua_isnil(L, -1);
+
+	if (!isCached)
+		lua_pop(L, 1); // remove found "nil" from stack
+
+	LUA_DEBUG_END(L, (isCached ? 1 : 0));
+	return isCached;
+}
+
+/**
+* Finds field in the cache.
+* Require a existing cache table in the stack at index -1.
+* Returns true if field not nil (exists), puts value to lua stack.
+*/
+static bool import_from_cache(lua_State *L, const ImportInfo& importInfo)
+{
+	LUA_DEBUG_START(L);
+	assert(lua_istable(L, -1));
+	DEBUG_INDENTED_PRINTF("import [%s]: trying to load from cache...", importInfo.importName.c_str());
+
+	bool isImported = false;
+	std::string cacheName;
+
+	size_t dirsToCheck = 1; // check by original name
+	if (!importInfo.isFullName)
+		dirsToCheck += importInfo.importDirectories.size(); // check also relative names
+
+	// check original name, with lua extension and relative by dirs by same rule
+	// if isFullName is true, loop pass once and check only original importName
+	for (size_t i = 0; i < dirsToCheck; i++)
+	{
+		if (i == 0) cacheName = importInfo.importName;
+		else cacheName = FileSystem::NormalisePath( // normalize for relative paths as "../target"
+			FileSystem::JoinPath(importInfo.importDirectories[i - 1], importInfo.importName));
+
+		// check original name
+		isImported = get_cached(L, cacheName);
+
+		// check name with extension
+		if (!isImported && !importInfo.isFullName)
+		{
+			cacheName += ".lua";
+			isImported = get_cached(L, cacheName);
+		}
+
+		if (isImported)
+		{
+			DEBUG_PRINTF(" found cached %s\n", cacheName.c_str());
+			break;
+		}
 	}
-	lua_pop(L, 1);
-	DEBUG_PRINTF("import: not cached %s\n", cacheName.c_str());
 
-	// next we're going to look in libs/
-	cacheName = FileSystem::JoinPath("libs", importName+".lua");
-	lua_getfield(L, -1, cacheName.c_str());
-	if (!lua_isnil(L, -1)) {
-		// found it, use it
-		DEBUG_PRINTF("import: found cached %s\n", cacheName.c_str());
-		lua_remove(L, -2);
-		return 1;
+	if (!isImported)
+		DEBUG_PRINTF(" not cached\n");
+
+	LUA_DEBUG_END(L, (isImported ? 1 : 0));
+	return isImported;
+}
+
+/**
+* Tries to load file.
+* Returns true if file exists and puts return of dofile to stack.
+*/
+static bool load_file(lua_State *L, const std::string& name)
+{
+	LUA_DEBUG_START(L);
+
+	bool loaded = false;
```

Comment:
Loaded is useless here, you can just check if ```fileData != null```

Reply:
@laarmen it is needed for that line:
~~~cpp
LUA_DEBUG_END(L, (loaded ? 1 : 0));
~~~

Reply:
LUA_DEBUG_END(L, (fileData ? 1 : 0)); works too.